remoteappmanager/handlers/home_handler.py

```diff
@@ -79,46 +54,74 @@ def post(self):
 
             # Render the home page again with the error message
             # User-facing error message (less info)
-            message = ('Failed to {action} "{image_name}". '
+            message = ('Failed to {action}. '
                        'Reason: {error_type} '
                        '(Ref: {ref})')
-            self.render('home.html', images_info=images_info,
+            self.render('home.html',
+                        images_info=images_info,
                         error_message=message.format(
                             action=action,
-                            image_name=options["image_name"][0],
                             error_type=type(e).__name__,
                             ref=ref))
 
     # Subhandling after post
 
     @gen.coroutine
-    def _actionhandler_start(self, user_name, options):
+    def _actionhandler_start(self, options):
         """Sub handling. Acts in response to a "start" request from
         the user."""
-        # Start the single-user server
+        container_manager = self.application.container_manager
+        orm_user = self.current_user.orm_user
+
+        mapping_id = options["mapping_id"][0]
+
+        session = self.application.db.create_session()
+
+        with orm.transaction(session):
+            choice = [(m_id, app, policy)
+                      for m_id, app, policy in orm.apps_for_user(session,
+                                                                 orm_user)
+                      if m_id == mapping_id]
+
+            if len(choice) == 0:
+                raise ValueError("User is not allowed to run the application.")
+
+        _, app, policy = choice[0]
+
+        container = None
 
         try:
-            # FIXME: too many operations in one try block, should be separated
-            # for better error handling
-            image_name = options["image_name"][0]
-            container = yield self._start_container(user_name, image_name)
+            container = yield self._start_container(orm_user,
+                                                    app,
+                                                    policy,
+                                                    mapping_id)
+
             yield self._wait_for_container_ready(container)
         except Exception as e:
             # Clean up, if the container is running
-            yield self._stop_and_remove_container(user_name, image_name)
+            if container is not None:
```

Comment:
Ok then, how can we clean up a container without knowing what it is...
The problem is in start_container.